// @apollo/space-kit/Table/Table.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var _tslib = require('../_tslib-bcbe0269.js');
var core = require('@emotion/core');
var React = require('react');
var React__default = _interopDefault(React);
require('../fonts/fonts.js');
var base$1 = require('../base-91f9e1df.js');
require('../mono-a0d4bd30.js');
var colors = require('../colors/colors.js');
var assertUnreachable = require('../assertUnreachable-698a501a.js');

/**
 * Take an `as` value and make it into a `React.ReactElement`
 */

function createElementFromAs(as) {
  return /*#__PURE__*/React__default.isValidElement(as) ? as : typeof as === "string" ? /*#__PURE__*/React__default.createElement(as) : assertUnreachable.assertUnreachable(as);
}
/**
 * Tables provide a structure to data and a visual grid making it easier to see
 * relationships and are one of the most useful tools and formats for organizing
 * and communiting structured data.
 *
 * @see https://zpl.io/bAlrjJe
 */


function Table(_a) {
  var _b = _a.as,
      as = _b === void 0 ? /*#__PURE__*/React__default.createElement("table", null) : _b,
      data = _a.data,
      _c = _a.density,
      density = _c === void 0 ? "standard" : _c,
      columns = _a.columns,
      keyOn = _a.keyOn,
      _d = _a.trAs,
      trAs = _d === void 0 ? "tr" : _d;
  var padding = density === "standard" ? 8 : density === "condensed" ? 3 : 11;
  var getRowKey = typeof keyOn === "function" ? keyOn : function (row) {
    return row[keyOn];
  };
  var headTrElement = /*#__PURE__*/React__default.isValidElement(trAs) ? trAs : typeof trAs === "string" ? /*#__PURE__*/React__default.createElement(trAs) : createElementFromAs(trAs.head || "tr");
  var bodyTrElement = /*#__PURE__*/React__default.isValidElement(trAs) ? trAs : typeof trAs === "string" ? /*#__PURE__*/React__default.createElement(trAs) : createElementFromAs(trAs.body || "tr");
  return /*#__PURE__*/React__default.createElement(core.ClassNames, null, function (_a) {
    var css = _a.css,
        cx = _a.cx;
    return /*#__PURE__*/React__default.cloneElement(as, {
      className: cx(css({
        borderCollapse: "collapse",
        width: "100%"
      }), as.props.className)
    }, /*#__PURE__*/React__default.createElement(React__default.Fragment, null, /*#__PURE__*/React__default.createElement("colgroup", null, columns.map(function (_a) {
      var colProps = _a.colProps,
          id = _a.id;
      return /*#__PURE__*/React__default.createElement("col", _tslib.__assign({
        key: id
      }, colProps));
    })), /*#__PURE__*/React__default.createElement("thead", null, /*#__PURE__*/React__default.cloneElement(headTrElement, {
      className: cx(css(_tslib.__assign(_tslib.__assign({}, base$1.xsmall), {
        borderBottom: "1px solid " + colors.colors.silver.dark,
        color: colors.colors.grey.darker,
        textAlign: "left",
        textTransform: "uppercase"
      })), headTrElement.props.className)
    }, /*#__PURE__*/React__default.createElement(React__default.Fragment, null, columns.map(function (_a, colIndex) {
      var headerTitle = _a.headerTitle,
          id = _a.id,
          _b = _a.thAs,
          thAs = _b === void 0 ? "th" : _b;
      var element = createElementFromAs(thAs);
      return /*#__PURE__*/React__default.cloneElement(element, {
        className: css(css({
          fontWeight: 600,
          padding: padding,
          paddingLeft: colIndex === 0 ? 0 : padding,
          paddingRight: colIndex === columns.length - 1 ? 0 : padding
        }), element.props.className),
        key: id
      }, headerTitle);
    })))), /*#__PURE__*/React__default.createElement("tbody", null, data.map(function (item, index) {
      return /*#__PURE__*/React__default.cloneElement(bodyTrElement, {
        key: getRowKey(item)
      }, /*#__PURE__*/React__default.createElement(React__default.Fragment, null, columns.map(function (_a, colIndex) {
        var _b = _a.as,
            as = _b === void 0 ? "td" : _b,
            render = _a.render,
            id = _a.id;
        var element = createElementFromAs(as);
        return /*#__PURE__*/React__default.cloneElement(element, {
          key: id,
          className: cx(css({
            // no border on the bottom row
            borderBottom: index === data.length - 1 ? "none" : "1px solid " + colors.colors.silver.dark,
            padding: padding,
            paddingLeft: colIndex === 0 ? 0 : padding,
            paddingRight: colIndex === columns.length - 1 ? 0 : padding
          }), element.props.className)
        }, render(item, index, data));
      })));
    }))));
  });
}

exports.Table = Table;
